front/lib/resources/vault_resource.ts

```diff
@@ -391,107 +392,95 @@ export class VaultResource extends BaseResource<VaultModel> {
     });
   }
 
+  private mapGroupPermissions(permissions: Permission[]) {
+    return this.groups.map((group) => ({
+      id: group.id,
+      permissions,
+    }));
+  }
+
   /**
-   * Determine ACL entries based on vault type and group configuration:
-   * 1. For regular vaults with a global group:
-   *    - Return only the global group with full permissions
-   *    - Ignore regular groups as they're not used in non-restricted vaults
-   * 2. For all the other vaults:
-   *    - Return all associated groups with full permissions
+   * Computes Access Control List (ACL) entries based on vault type and group configuration.
+   *
+   * Permission patterns:
+   * - System vaults: admin-only write access
+   * - Public vaults: anyone can read, workspace members can write
+   * - Global vaults: admin and builder have read/write access
+   * - Regular vaults with global group: hierarchical access (admin > builder > user)
+   * - Restricted vaults: admin access + explicit group permissions
    */
   acl(): ACLType {
     const globalGroup = this.isRegular()
       ? this.groups.find((group) => group.isGlobal())
       : undefined;
+
+    if (this.isSystem()) {
+      return {
+        workspaceId: this.workspaceId,
+        roles: [{ name: "admin", permissions: ["write"] }],
+        groups: [],
+      };
+    }
+
+    if (this.isPublic()) {
+      return {
+        workspaceId: this.workspaceId,
+        roles: [
+          { name: "admin", permissions: ["admin", "read", "write"] },
+          { name: "builder", permissions: ["read", "write"] },
+          { name: "user", permissions: ["read", "write"] },
+          // Everyone can read.
+          { name: "none", permissions: ["read"] },
+        ],
+        groups: this.mapGroupPermissions(["read", "write"]),
+      };
+    }
+
+    if (this.isGlobal()) {
+      return {
+        workspaceId: this.workspaceId,
+        roles: [
+          { name: "admin", permissions: ["read", "write"] },
+          { name: "builder", permissions: ["read", "write"] },
+        ],
+        groups: this.mapGroupPermissions(["read"]),
```

Comment:
I would inline (for readability)

Reply:
inline?

Reply:
move the this.groups.map(...) instead of calling a proxy function. (i know you are doing it 3 times but it's one liner)